seed + serialize LessonActivity and ActivitySection

davidsbailey opened the pull request:
## Background

Finishes [PLAT-506](https://codedotorg.atlassian.net/browse/PLAT-506). For background on how the new seeding system works, see:
* the "Strategy" section at https://github.com/code-dot-org/code-dot-org/pull/36733 
* code comments in https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/lib/services/script_seed.rb

Currently, we use the new serialize / seed code in the following places:
* in levelbuilder mode, when updating a script (legacy or migrated), the .script_json is updated in addition to the .script file.
* when seeding in any environment, the script level seeding_key is generated. this is a step toward launching new seeding for existing scripts, which we do not plan to do until after April.

For January launch, the general idea is to enable new script seeding for _migrated scripts only_. This will allow us to restore a previous version of a 21-22 script from a previous point in the github history if we need to, and will help us prove that serialization and seeding are correctly propagating changes to other environments. Meanwhile, we'll defer the high cost of launching new serialization + seeding for legacy scripts until after April launch.

## Description

What this PR does:
1. **adds LessonActivity and ActivitySection to the new serialize + seed codepaths**.  Fortunately, Winter left some excellent instructions on how to add a new model: https://github.com/code-dot-org/code-dot-org/blob/b336429a6d2ea5d71e9d2b3478ca61d2dd439e8a/dashboard/test/lib/services/script_seed_test.rb#L4-L14 Individual commits are named to match up with the items in this list.  

2. **adds the ScriptLevel.activity_section relationship to serialize + seed**. This is done in f33b7cda705b5f042ebb475f948c81f9ac565ed6 and parts of 00ed5ddd27f5e0f068671b53a79be4b97c1c003b . 

3. the rest of the commits at the end of the list are various cleanup and optimizations.

## Testing story

Added new unit tests per the checklist in the description. Some of these are really integration tests, because they do a "round trip" from DB objects --> json --> DB objects. There are no end-to-end tests yet because we are not using this code end-to-end anywhere in the product yet.

## Future work

* make ScriptSeed autoload / reload. This doesn't work right now because script_seed.rb is required explicitly. Without this, after making changes to script_seed.rb (such as if you pull down this PR locally), you have to run `spring stop` before re-running `spring testunit`, and you have to kill and rerun bin/dashboard-console or bin/dashboard-server. This makes development a bit slower and a lot more error-prone. The solution is to refer to it as Services::ScriptSeed so that rails can find and autoload it, and then not require it explicitly. This may touch a lot of lines, so I'll do this in the next PR.

* The rest of the tasks from the [PLAT-360](https://codedotorg.atlassian.net/browse/PLAT-360) epic which are in sprints X or Y

* `ScriptSeedTest.test_seed_new_script` fails when run individually, due to 2 unidentified additional queries when compared with running the entire test file. (won't fix for now)

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

bethanyaconnor commented on dashboard/app/models/script_level.rb:
```diff
@@ -117,7 +117,7 @@ def self.add_script_levels(script, lesson_group, lesson, raw_script_levels, coun
       # Generate and store the seed_key, a unique identifier for this script level which should be stable across environments.
       # We'll use this in our new, JSON-based seeding process.
-      seed_context = ScriptSeed::SeedContext.new(script: script, lesson_groups: [lesson_group], lessons: [lesson])
+      seed_context = ScriptSeed::SeedContext.new(script: script, lesson_groups: [lesson_group], lessons: [lesson], lesson_activities: [], activity_sections: [])
```
Shouldn't some script levels have activity sections?

davidsbailey replied:
hmm... yes... I was thinking this would only be done for legacy scripts, because it is part of the old seed process. but since we are still using the old seed process on migrated scripts, this will have to change. 

I was thinking of just removing this block of code, since it is only needed as a sort of dry run for launching new script seeding for legacy scripts. but it seems like we might not be trying to do that any time soon, or we may never do it in the sense that we may just "migrate" these scripts instead, which would be a different way to get them onto the new script seeding system. How does that sound?

davidsbailey replied:
I went ahead and removed this block of code in this PR, since I don't think we want to risk leaving things in a state where the old/current seed process running could mess up imported/migrated scripts.

davidsbailey replied:
removing this broke some tests, most notably this makes it so that we don't have anything preventing us from adding the same level to a script twice:
```
===[0m[1000D[K FAIL["test_level_can_only_be_added_once_to_a_lesson", "ScriptTest", 969.5705375440011]
 test_level_can_only_be_added_once_to_a_lesson#ScriptTest (969.57s)
        StandardError expected but nothing was raised.
        test/models/script_test.rb:2656:in `block in <class:ScriptTest>'
        test/testing/setup_all_and_teardown_all.rb:36:in `run'

[0m[1000D[K FAIL["test_script_import_is_idempotent_w.r.t._positions_and_count", "ScriptTest", 971.1916625019949]
 test_script_import_is_idempotent_w.r.t._positions_and_count#ScriptTest (971.19s)
        Expected: 5
          Actual: 0
        test/models/script_test.rb:310:in `block in <class:ScriptTest>'
        test/testing/setup_all_and_teardown_all.rb:36:in `run'
```
I need to think on it a bit more to figure out what to do here.

davidsbailey replied:
I've restored this block of code for now. I see a few paths forward and want to get my thoughts down about them.

this codepath serves two purposes:

https://github.com/code-dot-org/code-dot-org/blob/bc68aeeb69fc37209ded01f10e682d1cde34b8a5/dashboard/app/models/script_level.rb#L118-L122

1. this codepath is a first step toward simplifying some logic in the serialization/seeding process. see this draft PR: https://github.com/code-dot-org/code-dot-org/pull/36985 .
 
    * to explain, the idea is to simplify the code by making script_level.seed_key behave more like lesson_activity.key, i.e. use one value to identify the object rather than a set of values. Because script_levels already exist with blank keys in other environments (unlike lesson activities), this step is needed to give those script levels the same key across environments. we have to do this non-destructively because (and only because) teacher feedbacks depend on script_level_ids. so, if we were to fix teacher feedbacks first and only cared about this goal, then this goal could be accomplished by using `SecureRandom.uuid` here instead of by calling `seeding_key`.

2. this codepath enforces that levels are unique within the script, using the unique constraint in the DB on the seed_key column. this was an unintended consequence of (1), however it is valuable because:
  
    a. the new seed process can't distinguish between script levels with identical `seeding_key`s, (i.e. duplicate levels within the same _activity section_).

    b. progress is stored on (script_id, level_id). if a level is in the same script twice, one bubble will turn green when the other does.

    c. there may be other hidden things in our system which assume no duplicate levels in a script, too.

Because fixing teacher feedbacks makes this code unnecessary to accomplish goal (1), and goal (1) is not strictly urgent, I am proposing we decide based on goal (2).

**Do we agree that levels should be unique within a given script?** 

If we do, then I propose we keep this behavior by continuing to _exclude activity sections_ from inputs to `seeding_key` here, and add a comment explaining. Counterintuitively, this gives a stronger unique constraint (no duplicate script levels within _script_) than including it does (no duplicate script levels within _activity section_).

Regardless of what we choose, there will be follow-up work to do, because we handle duplicate levels differently in different places:
* **old seed** (this codepath) currently raises on duplicate levels within a script
* **new seed** silently overwrites the first script level if an identical one is found later in the _activity section_ (need to confirm)
* **lesson save** allows duplicate levels 
* **script save** (new and old) overwrites the first script level if an identical one is found later in the _script_

The code in this PR should now reflect this proposal. @dmcavoy @bethanyaconnor sorry for the long read. please let me know what you think. If this makes sense / sounds good, I'll add comments here explaining and we can discuss creating follow-up tasks to properly handle duplicate levels in other situations.

dmcavoy replied:
I think that the constraint of a level should be unique in a script makes sense to me

bethanyaconnor replied:
^ agreed. When I wrote my original comment, I didn't realize this was the old seeding logic 🙈. So this makes a lot of sense and I think it's fine to add the constraint of a level being unique within a script.

Are there any stakeholders we should ask before adding this hard constraint? I think it's been kind of an assumption before but if it's going to be codified in our seeding logic, it seems like it would be hard to change in the future

davidsbailey replied:
Thanks Bethany. General thoughts:
1. this PR doesn't add this constraint. it was added in October: https://github.com/code-dot-org/code-dot-org/pull/36984
2. when this constraint was added, there were only a couple of scripts which had duplicate levels. IIRC ~at most one~ none of them were in live scripts. I believe they were all cloned and/or removed in this commit: https://github.com/code-dot-org/code-dot-org/commit/1cb2d28f4c33eb4c01a54acfb23ce99c9dfb3acf 

Thinking through our new lesson edit UI, I would think that for now we could show an error if they try to save a lesson that contains a duplicate level. that might be all we need to do for now (January Launch? January ASAP?) to keep bad data from getting into our system. Opened here: https://codedotorg.atlassian.net/browse/PLAT-612

 in the future, I could see graying out the `+` button in the add level dialog to make this a bit more user-friendly. I'm not sure if this needs to be tracked -- we could also wait and see if anyone runs into this before opening a ticket for it.

davidsbailey replied:
@bethanyaconnor you make a good point about considering stakeholders needs as we commit harder in this direction. I think our long term vision is to get closer to levels appearing in only one place, so I think this is something we actually want to push for as an engineering team. so I am a bit hesitant to ask them if they want this. if they did have a need for it, I don't think there is actually anything stopping us from adding, say, to the script_level's seeding_key to enable this. so that makes me want to just not support it now, and add it later if we need it.

bethanyaconnor replied:
That all sounds good to me!

davidsbailey replied:
ok, great! The Jira item for the follow up which I think we should really do soon is linked above, and I've updated the comment here to reflect the two purposes of seed_key.

bethanyaconnor commented on dashboard/app/models/activity_section.rb:
```diff
@@ -86,6 +87,23 @@ def update_script_levels(script_levels_data)
+    raise "No LessonActivity found for #{self.class}: #{my_key}, LessonActivity ID: #{lesson_activity_id}" unless my_lesson_activity
+    {
+      'activity_section.key': key,
+      'lesson_activity.key': my_lesson_activity.key
```
Question for my own benefit: why do we need to include `lesson_activity.key` here? Shouldn't `activity_section.key` be enough to uniquely identify an activity section?

davidsbailey replied:
Great question, which gets at the core of the new seeding strategy. The seeding_key must have enough information to uniquely identify the object within the script, but it also needs to tell us where to put the object (this part may be missing from the comments). this object is NOT nested within its parent (lesson_activity) in the serialized format, so we still need the `lesson_activity_key` in order to know where to put it.

With `m` models and `n` objects/instances of those models, this allows us to do `O(m)` DB queries to import all the data into each script, no matter how many objects there are. With a nested structure, the number of queries would increase with `n`.

bethanyaconnor commented on dashboard/app/models/script_level.rb:
```diff
@@ -623,8 +623,17 @@ def seeding_key(seed_context, use_existing_level_keys = true)
+    # going to be used for seeding yet.
+    my_activity_section = seed_context.activity_sections.select {|s| s.id == activity_section_id}.first
+
+    # HACK: use script_level prefix so that it gets filtered out later.
```
I think I'm just missing it but what do you mean by "script_level prefix"?

davidsbailey replied:
oops -- I had called this `'script_level.activity_section.key'` at first, which was a hack to take advantage of this hack: https://github.com/code-dot-org/code-dot-org/blob/9705dd7988b0a4025212089f7385f7aecc6fdb5d/dashboard/lib/services/script_seed.rb#L183-L185 I fixed the hack and squashed the commits, but forgot to remove this comment.

davidsbailey replied:
also fixed in this PR. the merge conflicts weren't too bad after all.

bethanyaconnor commented on dashboard/lib/services/script_seed.rb:
```diff
@@ -175,22 +186,63 @@ def self.import_lessons(lessons_data, seed_context)
 
+  def self.import_lesson_activities(activities_data, seed_context)
+    activities_to_import = activities_data.map do |activity_data|
+      lesson_id = seed_context.lessons.select {|l| l.key == activity_data['seeding_key']['lesson.key']}.first&.id
```
Should there be a check that this lesson actually exists? Or is that what `seed_context` is for?

davidsbailey replied:
yep, that's what seed_context is for. it prevents us from having to go to the DB to see if the lesson exists.

bethanyaconnor replied:
Make sense -- thanks for the explanation!

bethanyaconnor commented on dashboard/test/lib/services/script_seed_test.rb:
```diff
@@ -138,6 +195,8 @@ class ScriptSeedTest < ActiveSupport::TestCase
     expected_counts = original_counts.clone
     expected_counts['LessonGroup'] -= 1
     expected_counts['Lesson'] -= 2
+    expected_counts['LessonActivity'] -= 2
```
Might be out of the scope of this PR but is there a test that the correct Lessons/LessonActivities/etc are getting deleted? This part of the code makes me super nervous

davidsbailey replied:
Great point. There is not. I will think about how to add this in a subsequent PR.

davidsbailey replied:
I don't think this is going to be quick, since I am having some trouble with deletion tests in a subsequent PR. I'm going to open a Jira task to make sure we don't lose track of this: https://codedotorg.atlassian.net/browse/PLAT-609

bethanyaconnor reviewed (commented):
I've been over this PR a couple of times and I think I have an idea of what's going on but I have a few basic questions

davidsbailey commented:
Good questions so far Bethany! Because the next PR touches many lines by adding indentation, I'm going to propose addressing feedback in a subsequent PR. I'll send that PR out before merging this one so we can see nothing has been overlooked.

davidsbailey commented on dashboard/app/models/activity_section.rb:
```diff
@@ -86,6 +87,23 @@ def update_script_levels(script_levels_data)
     end
   end
 
+  # Used for seeding from JSON. Returns the full set of information needed to uniquely identify this object.
```
would this be clearer @bethanyaconnor ?
```suggestion
# Used for seeding from JSON. Returns the full set of information needed to uniquely identify this object as well as any other objects it refers to by id. 
```
double checking ahead of time since this comment is duplicated in many places.

bethanyaconnor replied:
Yes, I think so. I think "uniquely identify this object" made me think that it was in the scope of only this object, and sometimes you need to know a relationship to uniquely identify it. 

small nit: I'm not sure `id` is right here? You use `key` in this particular function for instance. Can the comment be simplified to "as well as a unique identifier for any other objects it refers to" or something like that?

davidsbailey replied:
please do nit! I struggled with how to word this, because I wanted something that gave the right idea but is also technically accurate and not too long.

conceptually, you need to be able to be able to identify the object, and put it into the right place in the hierarchy. technically, you are filling all the fields named `id` or `*-id`. Saying "refers to" gives the right idea, but is a bit ambiguous because it makes me think of associations in both directions.

What do you think about **"as well as a unique identifier for any other objects it belongs to"**? That has both the right idea (objects belong to their parents) and implies a technically accurate description (`belongs_to` implies that this object has a field which refers to another object by id).

bethanyaconnor replied:
Yeah I like that!

davidsbailey reviewed (commented):
@bethanyaconnor I ended up addressing your comments in this PR, except for the one thing I opened a new Jira item for.

dmcavoy commented on dashboard/app/models/script_level.rb:
```diff
@@ -115,12 +115,6 @@ def self.add_script_levels(script, lesson_group, lesson, raw_script_levels, coun
         sl.levels = levels
       end
 
-      # Generate and store the seed_key, a unique identifier for this script level which should be stable across environments.
```
Why did you remove this?

davidsbailey replied:
I'm not sure yet if I should actually remove this. please see https://github.com/code-dot-org/code-dot-org/pull/38145#discussion_r539854584

davidsbailey replied:
sorry, I meant https://github.com/code-dot-org/code-dot-org/pull/38145#discussion_r537936192 .

dmcavoy commented on dashboard/test/fixtures/test-serialize-seeding-json.script_json:
```diff
@@ -101,10 +101,111 @@
+  "lesson_activities": [
+    {
+      "key": "test-serialize-seeding-json-lg-1-lesson-1-activity-1",
+      "position": 1,
```
it looks like position is always 1. is that correct?

davidsbailey replied:
I took a shortcut in `create_script_tree` and only made there be 1 LessonActivity per Lesson, and only 1 ActivitySection per LessonActivity. So, this is correct in one sense (working as I currently intended), but not in another (Winter would probably say that there should always be at least 2 things in a list in tests). Since this PR is getting so big, and nothing is relying on this code yet, how about I work on increasing the number of these from 1 to 2 in a separate PR? I'll add this as a note in [PLAT-609](https://codedotorg.atlassian.net/browse/PLAT-609).

dmcavoy replied:
Waiting until a later PR sounds fine to me

bethanyaconnor reviewed (commented):
Thanks for answering all my questions! I have one more: where do we decide which seeding process to use for an individual script?

davidsbailey commented:
> Thanks for answering all my questions! I have one more: where do we decide which seeding process to use for an individual script?

I think this is captured here: https://codedotorg.atlassian.net/browse/PLAT-560 "launch new serialize + seed for migrated scripts (LB only)" I just went and added some ideas for how to implement this.

bethanyaconnor reviewed (approved):
Oh perfect thanks!

davidsbailey commented:
Thanks for thinking through this with me, Bethany!
